### ISSUE TITLE
Renaming / moving stuff around

### ISSUE THREAD

**Issue by dmp42** [docker/distribution#225]
It is a regularly requested feature to be able to "rename" a repository. Arguably, that can be done by locally renaming the image and pushing it again, but that can be tedious if there are many tags.

Also, renaming any random part of the full image name has been requested as well (eg: "change my username to X and have all images follow that").

Doing this "à-la-github" would in addition require to put redirects in place.

**Comment by dmp42** [comment]
See for example https://github.com/docker/docker-registry/issues/950

**Comment by endophage** [comment]
Would the owner be able to re-use the old name? It seems dangerous as it would let them pull a bait and switch, but would also be really annoying for some users to prohibit reuse.

**Comment by dmp42** [comment]
@endophage agreed that repo rename and account rename are probably two different beasts.
Agreed that renaming user accounts is probably more of a problem.
Repo rename should probably add redirects from the old repo to the new and leave the former "unusable".

Invoking our favorite @metalivedev @shin- and @kencochrane for some input here, since they are first line.

**Comment by stevvooe** [comment]
Should this be a part of #52?

**Comment by dmp42** [comment]
@stevvooe maybe - though if we decide to have some redirection (a-la-github) for renamed repos, that might exceed the scope of it.

**Comment by stevvooe** [comment]
@dmp42 Aliasing? That might not that unrealistic, although we need to be a little careful with how that interacts with the docker daemon.

**Comment by kencochrane** [comment]
Ideally it would be nice to handle both namespace rename as well as repo rename. But if we can only get one of the two, that is a good start, and better then nothing :) 

Redirection would also be preferred, but like @stevvooe said we would need to be careful for how it interacts with docker engine. I'm not sure if it is best to leave the old name "unusable" or allow the ability for people to reuse that name. I think it would be nice to allow them to reuse it, but I can see how this could cause some security concerns with a bait and switch. Github allows you to reuse repo names after you rename them, as an example.

**Comment by shin-** [comment]
I know it's been a requested feature for a long time, but v1 makes this complicated because of the overall way it was architected. I don't have insight on how v2 handles this. Assuming v2 makes this somewhat simple, can we afford UI/UX wise to only make this available to v2 repos? The hub part of the feature is relatively trivial.

**Comment by stevvooe** [comment]
@shin- @kencochrane The repository is relatively self-contained. This would be as easy as moving the underlying file path. It seems logical that this would be a component of the tool described in #52. I see no problem adding support for the following:

1. Add management tool functionality to rename a repository.
2. Add management tool functionality to alias one repository to another.

These would give registry users the ability to move repositories and make policy decisions about redirects depending on the use case.

Part of the issue here is that we have enforced that a repository name has to match the image name. This causes a lot of problems with moving images, signing and verification that may not be worth it.

**Comment by kencochrane** [comment]
@stevvooe So is the fact that we enforced repo names to match image names, a show stopper, and we can't add this functionality rename/redirect functionality?

**Comment by stevvooe** [comment]
@kencochrane It's not a show stopper. The registry should support this use case but possibly block serving up tags with a different repository name, if present.

Ideally, we separate naming from transport. If we can do that, then matching the repository name with the image name is only a convention. Moving a repository can be a smooth and secure operation, where image names are only updated at the maintainers leisure.

For instance, let's say we rename "foo/bar" to "stevvooe/bar". The content immediately after would be something like this:

```
foo/bar:sometag -> <image A>
```

This could only be fetched as "stevvooe/bar:sometag", after a redirect from "foo/bar:sometag" but is verified as "foo/bar:sometag". At my leisure, I can name and sign a new tag without changing the target image:

```
foo/bar:sometag -> <image A>
stevvooe/bar:sometag -> <image A>
```

Now, it can be fetched and verified as "stevvooe/bar:sometag", replacing the value of "foo/bar:sometag". The redirect still works.

Much of the above is just a thought exercise but this is a good idea.

**Comment by xiekeyang** [comment]
**concept**

May rename the old one's Path and its manifest.
in source code, should involve that,
-[x] Add new routeDescriptors of /v2/<name> in /v2/descriptors.go
-[x] Construct new `func Rename` method in repository/middleware.go, via PUT method.
-[x] Add `func Update` in manifest methods.
-[x] Construct new API method `PATCH` the objective.

Current context use placeholder of {name:"newname"} as input name parameter. 
We select it, to be parsed via `RepositoryNameComponentRegexp` in descriptors.

If it is assigned and in implementation progress? 
Or I would like to put careful research.
Thanks a lot!

**Comment by stevvooe** [comment]
@xiekeyang This will not be added as an API method. Adding a `/v2/<name>` direct route will cause conflicts with existing routes. The possibility for abuse is too great.

This functionality will likely be part of a backend management tool or protocol. Please see #52 for details. I'd accept proposals along these lines that cover some or all of the use cases.

**Comment by cliffbdf** [comment]
Renaming is one thing, but how does one create a new repository to begin with? The v2 API is not clear on this. Someone in one thread commented that if one uploads a layer (blog), that the required repo will be created, but if I try that, I get a 404 Not Found. The API spec also says that one should first upload the layers, and then the image manifest - but it does not actually _say_ that the repo will be created if it does not exist. Should it say that? Note: I looked through all of the issues and this issue was the closest - I did not want to create a new issue until I was sure that this is actually a valid issue. (Related: the API seems to be missing functions for repo name management - listing names (which v2 allows to be hierarchical), setting ownership, etc.) Thanks!

**Comment by stevvooe** [comment]
@cliffbdf Repositories are created lazily when you upload to them. The specification does not mention the creation of repositories because repositories are just a partitioning convention, using a name. They are not created or destroyed. They just exist. It is their resources that are accessible.

Please open a new issue, describe _exactly_ what you are trying to do, along with the steps you took. Include what went wrong and what you expected to happen.

**Comment by cliffbdf** [comment]
@stevvooe Created issue, https://github.com/docker/distribution/issues/1711 Thanks!
